openmls/src/messages/external_proposals.rs

```diff
@@ -0,0 +1,84 @@
+//! External Proposal
+//!
+//! Contains the types and methods to build external proposal
+//! to add a client from a MLS group
+//! `Remove` & `ReInit are nto yet implemented`
+
+use crate::{
+    credentials::CredentialBundle,
+    error::LibraryError,
+    framing::{MlsMessageOut, MlsPlaintext, Sender},
+    group::{mls_group::errors::ProposeAddMemberError, GroupEpoch, GroupId},
+    key_packages::KeyPackage,
+    messages::{AddProposal, Proposal},
+};
+use openmls_traits::OpenMlsCryptoProvider;
+use tls_codec::TlsByteVecU8;
+
+/// External Proposal.
+/// External proposal allows parties outside a group to request changes to the latter.
+///
+/// This `enum` contains the different external proposals in its variants.
+/// Not yet implemented: `Remove` & `ReInit`
+#[non_exhaustive]
+pub enum ExternalProposal {
+    /// Proposes adding a client to a group.
+    /// Newly added client can be either the sender itself or not.
+    Add(AddProposal),
```

Comment:
While we leave out preconfigured senders, we can still think about what the API should look like eventually. The spec is a bit annoying because there is some overlap between an Add proposal from a preconfigured sender (who will by design always remain an external sender) and an Add proposal from a sender of type NewMember. 

I'm wondering if we should really treat both as external proposals, risking confusing users of the library. I see NewMember Add proposals more of yet another way to join the group (we already have Welcome and ExternalCommit).

Maybe this should be renamed to `JoinProposal` to avoid all confusion. 

I'm also not sure if we should have this `enum` to cover all proposals, because neither NewMember Add proposals nor ReInit proposals are really external proposals. Did you see a commonality between them I missed?

Reply:
> Maybe this should be renamed to JoinProposal to avoid all confusion

I agree with you on this one, this will make the API clearer.

> neither NewMember Add proposals nor ReInit proposals are really external proposals

I don't follow you on this one. I understood external proposal as "sent by a non-member". IMHO, listing all the proposals under this enum makes the API straightforward for someone learning MLS from the draft. One can jump back-and-forth between the draft and the code ; it's simpler than if it were scattered across the codebase.

Reply:
This part is complicated. 

These are the valid sender types for the various kinds of proposals where more than one sender type is allowed:

ReInit: `SenderType.member || SenderType.external`
Add: `SenderType.member || SenderType.external || SenderType.new_member_proposal`
Remove: `SenderType.member || SenderType.external`

Whenever the sender type is `external`, the group MUST use the `ExternalSender` extension. Note that this is not the case for `new_member_proposal`. This is where I see a big difference between "external" proposals and the JoinProposal.

We should use Rust's type safety to make it as easy and safe for consumers of the library, given how confusing this stuff is even for us.

I'd like to suggest that we limit this to an opinionated `JoinProposal` struct that has a `new(...)` method to safely create it. I don't see where the `ExternalProposal` enum is used anyway (unless I missed something). And we can revisit that when we introduce external/preconfigured senders.

On a similar note, we also need to think about what the inspection part of the API looks like. I think that part is missing so far. I'm thinking the easiest way would be to extend `StagedCommit` with a new category

Reply:
> I'd like to suggest that we limit this to an opinionated JoinProposal struct that has a new(...) method

Done. However I really wonder how will the API look like for `Preconfigured` senders since anyway you can't tie them to a `MlsGroup` to verify the `RequiredCapabilities` & `ExternalSender` extension. But as you said we can revisit that later

> we also need to think about what the inspection part of the API looks like

I introduced `ProcessedMessage::ExternalJoinProposalMessage` for that. It will later have to also encompass external commits since applications might define a slightly different authorization policy for each (at least Wire does)